### PR TITLE
Allow maps of repeated items, add some compile options

### PR DESCRIPTION
**Pull request by drmaas**
* option to disable rpcs
* option to prefix enums

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -14,8 +14,10 @@ go get -u github.com/NYTimes/openapi2proto/cmd/openapi2proto
 There are 4 CLI flags for using the tool:
 * `-spec` to point to the appropriate OpenAPI spec file
 * `-annotate` to include (google.api.http options) for [grpc-gateway](https://github.com/gengo/grpc-gateway) users. This is disabled by default.
-* `-out` to have the output written to a file rather than `Stdout. Defaults to `Stdout` if this is not specified`
+* `-out` to have the output written to a file rather than `Stdout`. Defaults to `Stdout` if this is not specified.
 * `-indent` to override the default indentation for Protobuf specs of 4 spaces.
+* `-generateRpcs=false` to disable generation of rpcs. This is enabled by default.
```
Comment (jprobinson) [inline README.md]: Do you have a a reason for not including the RPC definitions?

`protoc` ignores them unless the gRPC plugin is invoked and they do help define how the protobuf messages interact with one another.

File: cmd/openapi2proto/main.go [hunk 2]
```diff
@@ -21,10 +21,12 @@ func main() {
 }
 
 func _main() error {
-	specPath := flag.String("spec", "../../spec.yaml", "location of the swagger spec file")
-	annotate := flag.Bool("annotate", false, "include (google.api.http) options for grpc-gateway")
+	specPath := flag.String("spec", "../../spec.yaml", "location of the swagger spec file. Defaults to ../../spec.yaml if not set")
```
Comment (jprobinson) [inline cmd/openapi2proto/main.go]: The stdlib actually tacks on the defaults for us by default, no need to add this here:

```
╰─➤  openapi2proto -h                                                                                                                                                                                                                                                                                                     2 ↵
Usage of openapi2proto:
  -annotate
    	include (google.api.http) options for grpc-gateway
  -indent int
    	number of spaces used for indentation (default 4)
  -out string
    	the file to output the result to. Defaults to stdout if not set
  -spec string
    	location of the swagger spec file (default "../../spec.yaml")
```

File: compiler/compiler.go [hunk 3]
```diff
@@ -421,6 +437,16 @@ func (c *compileCtx) getBoxedType(t protobuf.Type) protobuf.Type {
 	}
 }
 
+func (c *compileCtx) getArrayType(name string) protobuf.Type {
```
Comment (jprobinson) [inline compiler/compiler.go]: This appears to return the same type no matter what. Is this switch even needed?

File: compiler/compiler.go [hunk 4]
```diff
@@ -391,8 +402,13 @@ func (c *compileCtx) getType(name string) (protobuf.Type, error) {
 		}
 
 		for t := range container {
-			if t.Name() == name {
-				return t, nil
+			if name == "array" {
+				c.addImportForType("google.protobuf.ListValue")
+				return c.getArrayType(name), nil
+			} else {
```
Comment (jprobinson) [inline compiler/compiler.go]: If this structure is needed, it'd be cleaner to do an `else if` or a `switch` statement here.

File: compiler/compiler.go [hunk 5]
```diff
@@ -391,8 +402,13 @@ func (c *compileCtx) getType(name string) (protobuf.Type, error) {
 		}
 
 		for t := range container {
-			if t.Name() == name {
-				return t, nil
+			if name == "array" {
+				c.addImportForType("google.protobuf.ListValue")
```
Comment (jprobinson) [inline compiler/compiler.go]: I need to read through this deeper, but is this adding `ListValue` as an import for _any_ nested array?

File: README.md (hunk 1 above)
Comment (drmaas) [reply README.md]: @jprobinson yes - we have use cases where we want to leverage protocol buffers outside of grpc, and so generating the rpc schemas and corresponding generated code will not be used. I agree this isn't a huge deal, but does help to remove clutter for our use case.

File: compiler/compiler.go (hunk 3 above)
Comment (drmaas) [reply compiler/compiler.go]: Nope, it isn't

**Comment by drmaas** [comment]
@jprobinson more background:

At Target we have specs that use maps, of which some have value types that are lists. Currently we are unable to transpile our specs with these maps into protobuf schemas. This provides a mechanism to provide the conversion. This does not change how non-map lists are described in the protobuf schema.

The rationale for the `generateRpcs` flag is more of a convenience to be able to generate schemas for use outside of grpc.

The rationale for `prefixEnums` is that our specs often have the same enum values in different schemas, but inside of the same package (the package is specified by the spec title, and since a single protobuf schema file with a single package is generated, the enum values are not namespaced). Adding a command line option for this allows us to transpile our specs.

**Comment by jprobinson** [comment]
Thanks for the explanation! I personally use this tool for non gRPC services so I can understand the want to clean up some of the RPC stuff in the proto spec.

Would you mind altering existing test fixtures or providing new ones to cover these scenarios?

File: README.md (hunk 1 above)
Comment (jprobinson) [reply README.md]: I'm not 100% sure why, but something feels off about a camel case flag. I also wonder if we can make it cleaner by somehow making it `false` by default so just including some flag would drop the RPCs.

How about something like `-no-rpc` or `-no-grpc`?

**Comment by drmaas** [comment]
@jprobinson yes I will do that. Also - I am looking into generating custom wrapper messages instead of `ListValue`. As you pointed out, it's untyped, which will be difficult to work with.

File: README.md [hunk 6]
```diff
@@ -14,8 +14,10 @@ go get -u github.com/NYTimes/openapi2proto/cmd/openapi2proto
...
+* `-prefixEnums` to enable inserting the enum name as an enum prefix for each value. This is disabled by default.
```
Comment (jprobinson) [inline README.md]: Ditto for the camel case here, maybe `-prefix-enums`? I also wonder if we can come up with a better name

**Comment by drmaas** [comment]
@jprobinson please re-review - the map lists are now statically typed and other suggestions have been implemented. I didn't add any tests yet, and can add those if desired. 

I ran this locally against a 6400 line spec, and compiled java code from the protobuf schema, so I am fairly certain it works as intended.

**Comment by drmaas** [comment]
Example of how this looks in protobuf:

```
message CartItemInsertRequest {
  ...
  string add_mode = 1;
  map<string, AssociatedPaymentsList> associated_payments = 2;
  ...
}
```

```
message AssociatedPaymentsList {
    repeated AssociatedPayment associated_payments = 1;
}
```

```
message AssociatedPayment {
    float amount = 1;
    string payment_id = 2;
}
```

File: compiler/compiler.go (hunk 5 above)
Comment (drmaas) [reply compiler/compiler.go]: This was removed as it is now using static list wrapper types.

File: README.md [hunk 7]
```diff
@@ -14,8 +14,10 @@ go get -u github.com/NYTimes/openapi2proto/cmd/openapi2proto
 There are 4 CLI flags for using the tool:
 * `-spec` to point to the appropriate OpenAPI spec file
 * `-annotate` to include (google.api.http options) for [grpc-gateway](https://github.com/gengo/grpc-gateway) users. This is disabled by default.
-* `-out` to have the output written to a file rather than `Stdout. Defaults to `Stdout` if this is not specified`
+* `-out` to have the output written to a file rather than `Stdout`. Defaults to `Stdout` if this is not specified.
 * `-indent` to override the default indentation for Protobuf specs of 4 spaces.
+* `-generate-rpcs=false` to disable generation of rpcs. This is enabled by default.
```
Comment (jprobinson) [inline README.md]: I think I'd still prefer to make this the opposite flag so it's a little cleaner from a CLI perspective.

For example, `-skip-rpc`(defaulted to false) versus `-generate-rpcs=false`. Thoughts?

File: compiler/compiler.go [hunk 8]
```diff
@@ -489,9 +507,33 @@ func (c *compileCtx) compileMap(name string, s *openapi.Schema) (protobuf.Type,
...
+			if v := s.Description; len(v) > 0 {
+				f.SetComment(v)
+			}
+			m.AddField(f)
+			m.SetComment("automatically generated wrapper for a list of " + baseFieldName + " items")
+			typ = m
+
+			// finally, make sure that this type is registered, if need be.
+			// hack alert - I couldn't figure out how to stop the wrappers from being specified more than once.
+			// listValueWrappers was my fix :(
```
Comment (jprobinson) [inline compiler/compiler.go]: Could we perhaps make this map a field on the `compileCtx`? Perhaps we could create a generic catcher of duplicate types and encapsulate it within `addTypeToParent`?

File: compiler/compiler.go [hunk 9]
```diff
@@ -555,7 +595,7 @@ func (c *compileCtx) compileSchema(name string, s *openapi.Schema) (protobuf.Typ
 	switch {
 	case s.Type.Empty() || s.Type.Contains("object"):
 		if ap := s.AdditionalProperties; ap != nil && !ap.IsNil() {
-			return c.compileMap(name, ap)
+			return c.compileMap(name, strings.TrimSuffix(rawName, "Message"), ap)
```
Comment (jprobinson) [inline compiler/compiler.go]: Is this in response to some internal styling you are parsing?

File: compiler/compiler.go [hunk 10]
```diff
@@ -833,6 +873,17 @@ func (c *compileCtx) parent() protobuf.Container {
 	return c.parents[l-1]
 }
 
+func (c *compileCtx) grandParent() protobuf.Container {
+	l := len(c.parents)
```
Comment (jprobinson) [inline compiler/compiler.go]: Mind putting this in a switch statement? It's a bit cleaner.

```go
switch len(c.parents) {
...
}
```

File: compiler/compiler.go [hunk 11]
```diff
@@ -421,6 +436,10 @@ func (c *compileCtx) getBoxedType(t protobuf.Type) protobuf.Type {
 	}
 }
 
+func (c *compileCtx) getArrayType() protobuf.Type {
```
Comment (jprobinson) [inline compiler/compiler.go]: This is just returning a constant so it's fine to inline above and just removing this method.

**Comment by jprobinson** [comment]
This is looking better, @drmaas, thanks!

I'd really appreciate it if you can include some testing for these updates you've added. I'm sure they work, but the tests will prevent future devs (like me 😅) from breaking the new functionality moving forward.

You can probably just tack on a couple maps in`fixtures/refs.(yaml|json)` and the expected results in `fixtures/refs.proto`

File: compiler/compiler.go (hunk 9 above)
Comment (drmaas) [reply compiler/compiler.go]: @jprobinson The variable name inside the wrapper was ending up with `Message` at the end, which does not align with the variable name in the input swagger spec.

**Comment by drmaas** [comment]
@jprobinson PTAL - tests added.

File: fixtures/refs.yaml [hunk 12]
```diff
@@ -15,3 +15,10 @@ definitions:
         additionalProperties:
           type: string
         x-proto-tag: 12
+      test_map_array:
```
Comment (jprobinson) [inline fixtures/refs.yaml]: Can we add one more test case that covers the generic `ListValue` type we've also added?

File: compiler/compiler.go (hunk 9 above)
Comment (jprobinson) [reply compiler/compiler.go]: Sounds fine. As long as the unit tests pass, we good.

**Comment by jprobinson** [comment]
🌯 ! Thank you for your work, @drmaas!

**Pull request merged**